### ISSUE TITLE
Viewport gets changed after window is made fullscreen

### ISSUE THREAD

**Issue by wouterbt** [glfw/glfw#2251]
In order to maintain a square OpenGL viewport when the window gets resized, I use a `framebufferSizeCallback` function to set the viewport. This works fine when the window is resized by dragging the corner, but when it is made fullscreen, the viewport is set correctly initially, but it is reset to the screen size after a couple of iterations of the main loop. I think this strange, undesired behaviour is due to a bug. The attached program illustrates this. It creates a window, then makes it full screen, then enters the main drawing and event handling loop. It counts the iterations of the main loop. Its output is as follows on my system:
```
Framebuffer size: 2880 x 1800
[0] Viewport set to (0, 0, 1800, 1800)
[15] Viewport set to (0, 0, 2880, 1800)
```
suggesting that the `framebufferSizeCallback` function is called and the viewport is set correctly initially, but gets changed after 15 iterations of the main loop. It does not matter whether the window is made fullscreen by the user (through the button in the title bar) or by the program (through ```glfwSetWindowMonitor```).

System details:  
MacBook Pro with MacOS 12.3.1  
GLFW version: 3.3.8 Cocoa NSGL EGL OSMesa dynamic (installed by Homebrew)  
Compiler: Apple clang version 13.1.6 (clang-1316.0.21.2.5), Target: x86_64-apple-darwin21.4.0  

Demonstration program `test_glfw.c`:
```c
// compile:
// cc -o test_glfw -lglfw -framework OpenGL -DGL_SILENCE_DEPRECATION test_glfw.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <GLFW/glfw3.h>
#include <OpenGL/GL.h>

// Make sure we always have a square viewport
void framebufferSizeCallback(GLFWwindow* window, int x, int y) {
    printf("Framebuffer size: %d x %d\n", x, y);
    if (x > y) {
        glViewport(0, 0, y, y);
    } else {
        glViewport(0, y - x, x, x);
    }
}

// Exit when a key is pressed
void keyCallback(GLFWwindow* window, int key, int scancode, int action, int mods) {
    glfwTerminate();
    exit(0);
}

int main(int argc, char** argv) {
    // Initialize the GLFW library
    if (!glfwInit()) {
        exit(-1);
    }
    
    // Report version
    printf("%s\n", glfwGetVersionString());

    // Create a windowed mode window and its OpenGL context
    GLFWwindow* window = glfwCreateWindow(200, 200, "Test Viewport", NULL, NULL);
    if (!window) {
        glfwTerminate();
        exit(-1);
    }

    // Make the window's context current
    glfwMakeContextCurrent(window);

    // Synchronize buffer swap interval with refresh rate
    glfwSwapInterval(1);

    // Set the callback function for resizing the window
    glfwSetFramebufferSizeCallback(window, framebufferSizeCallback);

    // Set the callback function for keypresses
    glfwSetKeyCallback(window, keyCallback);

    // Make the window full-screen
    GLFWmonitor *monitor = glfwGetPrimaryMonitor();
    const GLFWvidmode *mode = glfwGetVideoMode(monitor);
    glfwSetWindowMonitor(window, monitor, 0, 0, mode->width, mode->height, mode->refreshRate);
  
    while (!glfwWindowShouldClose(window)) {
        // Check if viewport has changed
        static int iteration = 0;
        static GLint previousViewport[4];
        GLint viewport[4];
        glGetIntegerv(GL_VIEWPORT, viewport);
        if (memcmp(previousViewport, viewport, sizeof(viewport))) {
            printf("[%d] Viewport set to (%d, %d, %d, %d)\n", iteration, viewport[0], viewport[1], viewport[2], viewport[3]);
            memcpy(previousViewport, viewport, sizeof(viewport));
        }
        iteration++;

        // Draw something
        glClear(GL_COLOR_BUFFER_BIT);
        glBegin(GL_LINES);
        glVertex2i(-1, -1);
        glVertex2i(1, 1);
        glVertex2i(-1, 1);
        glVertex2i(1, -1);
        glEnd();
        glfwSwapBuffers(window);

        glfwPollEvents();
    }
    
    glfwTerminate();
    return 0;
}
```

**Comment by ws909** [comment]
[Original post made on the forum](https://discourse.glfw.org/t/glviewport-not-working-with-maximized-window/2271/8)

The number of iterations in the loop seems to be irrelevant.

This issue is probably caused by the native frame buffer being resized after the callback is fired.


It's also worth mentioning that the frame buffer size callback is called before the normal size callback, which can lead to unnecessary computations if the normal size callback overrides the set size. Not directly related to this issue, but something that I discovered now.

**Comment by ws909** [comment]
@wouterbt Does this bug occur on Macs without the bus pocket on top of the screen?

**Comment by wouterbt** [comment]
> @wouterbt Does this bug occur on Macs without the bus pocket on top of the screen?

Eh? Bus pocket?

**Comment by ws909** [comment]
> Eh? Bus pocket?

Yes, the newer MacBook Pros have the same tiny rectangle in the top center of the screen, as the iPhones, where the camera is.

**Comment by wouterbt** [comment]
> Yes, the newer MacBook Pros have the same tiny rectangle in the top center of the screen, as the iPhones, where the camera is.

Alright, but I have never heard this called a bus pocket.
![bus-pocket](https://user-images.githubusercontent.com/20640938/215437091-f1499dcc-02d3-466a-a134-e9253f77df2d.jpg)

I do not have access to a Mac with this feature, so I can't say whether the bug occurs there.

**Comment by ws909** [comment]
@wouterbt Well, you answered my question. ;P

**Comment by brettyhale** [comment]
Unfortunately, this is a long standing 'bug' in `NSOpenGLContext` - see: [#80](https://github.com/glfw/glfw/issues/80)
The only solution that has worked consistently for me is to call `glViewport` in each redraw, which is portable, but *potentially* expensive. It also occurs if setting the viewport to a subset of the framebuffer on `FramebufferSize` callbacks, not just fullscreen modes, at *least* as recently as Catalina. When the redraw occurs, the viewport is reset - and not always immediately...

**Comment by dougbinks** [comment]
This does indeed look like #80.

As per the linked issue GLFW doesn't call `glViewport` itself.
